File: src/api/tasks.ts
import { invoke } from "@tauri-apps/api/core";

/** 任务状态 */
export type TaskStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

/** 任务类型 */
export type TaskType =
  | "embed_element"
  | "embed_chapter"
  | "sync_embeddings"
  | "generate_summary"
  | "generate_snapshots";

/** 异步任务实体（与后端 AsyncTask 一一对应） */
export interface AsyncTask {
  id: string;
  task_type: TaskType;
  project_id: string;
  target_type: string | null;
  target_id: string | null;
  content_hash: string | null;
  payload_json: string | null;

  status: TaskStatus;
  progress_current: number;
  progress_total: number;

  result_json: string | null;
  error_message: string | null;

  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/** 提交新任务的请求 */
export interface NewTaskRequest {
  task_type: TaskType;
  project_id: string;
  target_type?: string | null;
  target_id?: string | null;
  content_hash?: string | null;
  payload_json?: Record<string, unknown> | null;
}

/** Tauri 事件 payload：任务进度更新 */
export interface TaskProgressEvent {
  task_id: string;
  status?: TaskStatus;
  progress_current: number;
  progress_total: number;
}

/** Tauri 事件 payload：任务完成 */
export interface TaskCompletedEvent {
  task_id: string;
  status: TaskStatus;
  result?: unknown;
  error?: string;
}

/** 提交一个新任务（幂等） */
export async function submitTask(
  input: NewTaskRequest
): Promise<AsyncTask> {
  return invoke<AsyncTask>("submit_task", { input });
}

/** 查询项目下的任务列表 */
export async function listTasks(
  projectId: string,
  statusFilter?: TaskStatus
): Promise<AsyncTask[]> {
  return invoke<AsyncTask[]>("list_tasks", {
    projectId,
    statusFilter: statusFilter ?? null,
  });
}

/** 查询单个任务 */
export async function getTask(taskId: string): Promise<AsyncTask> {
  return invoke<AsyncTask>("get_task", { taskId });
}

/** 取消任务（仅 pending/running 可取消） */
export async function cancelTask(taskId: string): Promise<void> {
  return invoke("cancel_task", { taskId });
}

File: src-tauri/src/commands/tasks.rs
use sqlx::SqlitePool;
use tauri::{AppHandle, Manager, State};

use crate::error::AppError;
use crate::models::async_task::{AsyncTask, NewTaskRequest};
use crate::services::task_manager::TaskManagerState;

#[tauri::command]
pub async fn submit_task(
    app: AppHandle,
    pool: State<'_, SqlitePool>,
    task_manager_state: State<'_, TaskManagerState>,
    mut input: NewTaskRequest,
) -> Result<AsyncTask, AppError> {
    // 注入公共上下文：项目级任务需要 app_data_dir，但前端没有
    if matches!(
        input.task_type.as_str(),
        "sync_embeddings" | "generate_summary"
    ) {
        let app_data_dir = app.path().app_data_dir().map_err(|e| {
            AppError::Internal(format!("Failed to get app data dir: {}", e))
        })?;
        let mut payload = input.payload_json.unwrap_or_else(|| serde_json::json!({}));
        payload["app_data_dir"] = serde_json::Value::String(app_data_dir.to_string_lossy().to_string());
        input.payload_json = Some(payload);
    }

    let manager = task_manager_state.read().await.clone()
        .ok_or_else(|| AppError::Internal("TaskManager not initialized".into()))?;
    
    let task_id = manager.submit(input.clone()).await?;
    
    // 返回提交的任务
    let task = crate::db::repo::async_task_repo::get_by_id(&pool, &task_id)
        .await?
        .ok_or_else(|| AppError::NotFound(format!("Task '{}' not found after creation", task_id)))?;
    Ok(task)
}

#[tauri::command]
pub async fn list_tasks(
    pool: State<'_, SqlitePool>,
    project_id: String,
    status_filter: Option<String>,
) -> Result<Vec<AsyncTask>, AppError> {
    crate::db::repo::async_task_repo::list_by_project(&pool, &project_id, status_filter.as_deref()).await
}

#[tauri::command]
pub async fn get_task(
    pool: State<'_, SqlitePool>,
    task_id: String,
) -> Result<AsyncTask, AppError> {
    crate::db::repo::async_task_repo::get_by_id(&pool, &task_id)
        .await?
        .ok_or_else(|| AppError::NotFound(format!("Task '{}' not found", task_id)))
}

#[tauri::command]
pub async fn cancel_task(
    pool: State<'_, SqlitePool>,
    task_id: String,
) -> Result<(), AppError> {
    crate::db::repo::async_task_repo::update_status(
        &pool,
        &task_id,
        crate::models::async_task::TaskStatus::Cancelled,
        None,
        None,
        None,
        Some("Cancelled by user"),
    )
    .await
}
